android/libraries/rib-base/src/main/java/com/badoo/ribs/android/recyclerview/Adapter.kt

```diff
@@ -0,0 +1,103 @@
+import io.reactivex.functions.Consumer
+import java.util.UUID
+
+internal class Adapter<T : Parcelable>(
+    private val hostingStrategy: RecyclerViewHost.HostingStrategy,
+    initialEntries: List<Entry<T>>? = null,
+    private val router: RecyclerViewHostRouter<T>
+) : RecyclerView.Adapter<Adapter.ViewHolder>(),
+    Consumer<RecyclerViewHostFeature.State<T>> {
+
+    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+        var configurationKey: ConfigurationKey? = null
+        var uuid: UUID? = null
+    }
+
+    private var items: List<Entry<T>> = initialEntries ?: emptyList()
+
+    override fun getItemCount(): Int =
+        items.size
+
+    override fun accept(state: RecyclerViewHostFeature.State<T>) {
+        items = state.items
+
+        when (state.lastCommand) {
+            null -> { /* No-op when restored from TimeCapsule or genuinely empty state */
+            }
+            is Input.Add -> {
+                eagerAdd(state.items.last())
+                notifyItemInserted(state.items.lastIndex)
+            }
+
+        }
+    }
+
+    private fun eagerAdd(entry: Entry<T>) {
+        if (hostingStrategy == EAGER) {
+            router.add(entry.configurationKey, Item(entry.uuid))
+        }
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder =
+        ViewHolder(
+            FrameLayout(parent.context).apply {
+                layoutParams = FrameLayout.LayoutParams(
+                    FrameLayout.LayoutParams.MATCH_PARENT,
+                    FrameLayout.LayoutParams.WRAP_CONTENT
+                )
+            }
+        )
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        val entry = items[position]
+        holder.configurationKey = entry.configurationKey
+        holder.uuid = entry.uuid
+
+    }
+
+    override fun onViewAttachedToWindow(holder: ViewHolder) {
+        super.onViewAttachedToWindow(holder)
+        val configurationKey = holder.configurationKey!! // at this point it should be bound
+        if (hostingStrategy == LAZY) {
+            router.add(configurationKey, Item(holder.uuid!!))
+        }
+        router.activate(configurationKey)
+        router.getNodes(configurationKey)!!.forEach { childNode ->
+            childNode.attachToView(holder.itemView as FrameLayout)
+        }
+    }
+
+    override fun onViewDetachedFromWindow(holder: ViewHolder) {
+        super.onViewDetachedFromWindow(holder)
+        val configurationKey = holder.configurationKey!! // at this point it should be bound
+        router.deactivate(configurationKey)
+        router.getNodes(configurationKey)!!.forEach { childNode ->
+            childNode.detachFromView()
+        }
+        if (hostingStrategy == LAZY) {
+            router.remove(holder.configurationKey!!)
+        }
+    }
+
+    internal fun onDestroy() {
+        items.forEach {
+            router.deactivate(it.configurationKey)
```

Comment:
You do the same a bit above (`deactivate` -> `forEach { detachFromView }` )
Maybe extract?

Reply:
Done.